dask/array/tests/test_array_core.py

```diff
@@ -1065,6 +1065,51 @@ def test_operators():
     assert_eq(a, +x)
 
 
+class OtherTypeUfuncNone:
+    __array_ufunc__ = None
+
+    def __radd__(self, other):
+        return "radd"
+
+
+class OtherTypeUfuncUndefinedHighPriority:
+    __array_priority__ = 1000
+
+    def __radd__(self, other):
+        return "radd"
+
+
+class OtherTypeUfuncUndefinedLowPriority:
+    __array_priority__ = 1
+
+    def __radd__(self, other):
+        return "radd"
+
+
+def test_binary_operator_delegation():
+    # Binary operator delegation in Dask should follow Numpy's:
+    # https://numpy.org/neps/nep-0013-ufunc-overrides.html#behavior-in-combination-with-python-s-binary-operations
+
+    x = from_array([1, 2, 3])
+    ufunc_none = OtherTypeUfuncNone()
+    ufunc_undefined_high_priority = OtherTypeUfuncUndefinedHighPriority()
+    ufunc_undefined_low_priority = OtherTypeUfuncUndefinedLowPriority()
+
+    # If `other.__array_ufunc__ is None`, delegates back to Python
+    # and therefore call reflected operator on `other`
+    assert_eq(x + ufunc_none, "radd")
+
+    # If the `__array_ufunc__` attribute is absent on other and
+    # `other.__array_priority__ > self.__array_priority__`, also delegates back
+    # to Python and therefore call reflected operator on `other`
+    assert_eq(x + ufunc_undefined_high_priority, "radd")
+
+    # If `other.__array_priority__ <= self.__array_priority__`, does not
+    # delegate (here it raises an error)
+    with pytest.raises(TypeError):
```

Comment:
Looks like this one raises

Reply:
Indeed…

I'm new contributing to Dask. I couldn't find any contributer guide, like how to setup the dev env or run the unit tests locally. Did I miss something?

Reply:
This section should help: https://docs.dask.org/en/stable/develop.html#development-environment

Reply:
That's useful thanks.

However, after installing the dependencies, I can't seem to be able to run the tests.

```
% py.test dask --verbose --doctest-modules
ERROR: while parsing the following warning configuration:

  ignore:.*:dask.tests.warning_aliases.RemovedIn20Warning

This error occurred:

Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1690, in parse_warning_filter
    category: Type[Warning] = _resolve_warning_category(category_)
  File "/opt/homebrew/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1728, in _resolve_warning_category
    m = __import__(module, None, None, [klass])
ModuleNotFoundError: No module named 'dask'
```

I'm sure this is an obvious mistake…

Reply:
Did you do ``pip install -e .`` after installing the environment?

Reply:
My local env now works :)

Quick question: to write the test correctly, it would be best to use mock classes. I don't think you have [pytest-mock](https://pypi.org/project/pytest-mock/) as a test dependency. Do you think it's worth adding it?

It would allow something like this:

```python
def test_binary_operator_delegation(mocker):
    # Binary operator delegation in Dask should follow Numpy's:
    # https://numpy.org/neps/nep-0013-ufunc-overrides.html#behavior-in-combination-with-python-s-binary-operations

    x = from_array([1, 2, 3])
    ufunc_none = mocker.Mock()
    ufunc_undefined_high_priority = mocker.Mock()
    ufunc_undefined_low_priority = mocker.Mock()

    # If `other.__array_ufunc__ is None`, delegates back to Python
    # and therefore call reflected operator on `other`
    x + ufunc_none
    ufunc_none.__radd__.assert_called_once()

    # If the `__array_ufunc__` attribute is absent on other and
    # `other.__array_priority__ > self.__array_priority__`, also delegates back
    # to Python and therefore call reflected operator on `other`
    x + ufunc_undefined_high_priority
    ufunc_undefined_high_priority.__radd__.assert_called_once()

    # If `other.__array_priority__ <= self.__array_priority__`, does not
    # delegate (here it raises an error)
    x + ufunc_undefined_low_priority
    ufunc_undefined_low_priority.__radd__.assert_not_called()
```

Reply:
I've added this mock dependency, and re-wrote the tests.

Most of them seem now to pass :)

Reply:
The dependency is a bit unfortunate, wouldn't defining very small classes not work?

Reply:
I think it would, but it would be a much less "clean way" to test these, as we would rely on some side effects of the calls. It's actually surprising that such a large and well maintained project does not use mocks -- but if you think this is completely out of scope, I'll switch back to small classes! Let me know!

Happy holiday season :)